### PR TITLE
Crossover/Mutation

### PR DESCRIPTION
**Pull request by MRebolle**
Changes in Individual::export_genotype

### CODE REVIEW

File: pyrevolve/config.py [hunk 1]
```diff
@@ -67,7 +67,7 @@ def str_to_address(v):
 
 parser.add_argument(
     '--manager',
-    default=None,
+    default="experiments/examples/manager_neat.py",
```
Comment (portaloffreedom) [inline pyrevolve/config.py]: this should stay None

File: .gitignore [hunk 2]
```diff
@@ -75,4 +75,7 @@ Pipfile.lock
 
 # Revolve related
 *.sdf
-*.urdf
\ No newline at end of file
+*.urdf
+
+thirdparty/
+*.DS_Store
```
Comment (portaloffreedom) [inline .gitignore]: just `.DS_Store`, without '*'

File: .gitignore [hunk 3]
```diff
@@ -75,4 +75,7 @@ Pipfile.lock
 
 # Revolve related
 *.sdf
-*.urdf
\ No newline at end of file
+*.urdf
+
+thirdparty/
```
Comment (portaloffreedom) [inline .gitignore]: thirdparty needs to be there, not ignored

File: experiments/examples/manager_neat.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+from pyrevolve.genotype.plasticoding.mutation.mutation import MutationConfig as plasticMutationConfig
+from pyrevolve.genotype.lsystem_neat.mutation import standard_mutation as lmutation
+
+from pyrevolve.util.supervisor.analyzer_queue import AnalyzerQueue
+from pyrevolve.util.supervisor.simulator_queue import SimulatorQueue
+from pyrevolve.custom_logging.logger import logger
+from pyrevolve.genotype.plasticoding.plasticoding import PlasticodingConfig
+from pyrevolve.genotype.lsystem_neat.lsystem_neat_genotype import LSystemCPGHyperNEATGenotype, LSystemCPGHyperNEATGenotypeConfig
+from pyrevolve.genotype.neat_brain_genome.neat_brain_genome import NeatBrainGenomeConfig
+#from pyrevolve.evolution.individual import Individual
```
Comment (portaloffreedom) [inline experiments/examples/manager_neat.py]: remove this commented line

File: pyrevolve/evolution/individual.py [hunk 5]
```diff
@@ -33,7 +33,7 @@ def id(self):
         return _id
 
     def export_genotype(self, folder):
-        self.genotype.export_genotype(f'{folder}/genotypes/genotype_{self.phenotype.id}.txt')
+        self.genotype._body_genome.export_genotype(f'{folder}/genotypes/genotype_{self.phenotype.id}.txt') #Change
```
Comment (portaloffreedom) [inline pyrevolve/evolution/individual.py]: Why is this necessary?

File: pyrevolve/genotype/lsystem_neat/crossover.py [hunk 6]
```diff
@@ -1,12 +1,71 @@
...
+                sel_parent = random.randint(0, 1)
+                grammar[letter[0]] = parents[sel_parent].grammar[letter[0]]
+
+        new_genotype = Plasticoding(lsystem_conf.plasticoding, 'tmp')
+        new_genotype.grammar = grammar
+        return new_genotype
+
+
+
+    #mother_body = parent_individuals[0].genotype._body_genome
```
Comment (portaloffreedom) [inline pyrevolve/genotype/lsystem_neat/crossover.py]: remove commented code

File: pyrevolve/genotype/neat_brain_genome/mutation.py [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
 
 def _mutation(genotype, baby_is_clone: bool, search_mode: multineat.SearchMode, genotype_conf):
-    new_genotype = genotype.clone()
+    new_genotype = genotype.genotype._brain_genome.clone()
```
Comment (portaloffreedom) [inline pyrevolve/genotype/neat_brain_genome/mutation.py]: no, this was correct before

File: pyrevolve/genotype/neat_brain_genome/crossover.py [hunk 8]
```diff
@@ -1,24 +1,45 @@
...
+                                   lsystem_conf.neat.rng,
+                                   lsystem_conf.neat.multineat_params
+                                   )
+    return new_genotype
+
+
+
+
+
+    #mother = parent_individuals[0]
```
Comment (portaloffreedom) [inline pyrevolve/genotype/neat_brain_genome/crossover.py]: remove this commented code

**Review by portaloffreedom — CHANGES_REQUESTED** [review]
There are a few changes to do :)

also, remove all .DS_Store files that got committed

File: pyrevolve/evolution/individual.py (hunk 5 above)
Comment (MRebolle) [reply pyrevolve/evolution/individual.py]: maybe it isn't. I removed this and changed lsystem_neat_genotype

Comment (portaloffreedom) [reply pyrevolve/evolution/individual.py]: thanks, like this now is compatible with the old l-system genome :)

File: pyrevolve/genotype/lsystem_neat/crossover.py [hunk 9]
```diff
@@ -1,12 +1,62 @@
...
+    new_body = new_child_body(parents_body_genotype, lsystem_conf, crossover_conf)
+    new_brain = NEATBrainCrossover(parents_brain_genotype, Neatconf, crossover_conf, lsystem_conf)
+
+    child_genotype._body_genome = new_body
+    child_genotype._brain_genome = new_brain
+
     return child_genotype
+
+
+def new_child_body(parents, lsystem_conf, crossover_conf):
```
Comment (portaloffreedom) [inline pyrevolve/genotype/lsystem_neat/crossover.py]: can we use the lsystem crossover function here, instead of a copy paste? It seems familiar code to me but maybe I'm wrong and there are some necessary changes.

File: pyrevolve/genotype/neat_brain_genome/crossover.py [hunk 10]
```diff
@@ -1,24 +1,32 @@
...
-                                 genotype_conf.rng,
-                                 genotype_conf.multineat_params)
-    return child_genotype
+    assert len(parents) == 2
+
+    crossover_attempt = random.uniform(0.0, 1.0)
+    if crossover_attempt > crossover_conf.crossover_prob:
+         new_genotype = parents[0].genotype._brain_genome
+    else:
+        mother = parents[0]
```
Comment (portaloffreedom) [inline pyrevolve/genotype/neat_brain_genome/crossover.py:24]: aren't parents of type individual? Should it not be `parents[0].genotype`?

File: pyrevolve/genotype/neat_brain_genome/crossover.py [hunk 11]
```diff
@@ -1,24 +1,32 @@
...
 class NEATCrossoverConf:
     def __init__(self):
         self.mate_average = True
         self.interspecies_crossover = True
 
 
-def standard_crossover(parent_individuals, genotype_conf, crossover_conf):
+def standard_crossover(parents, NeatCrossoverConf, crossover_conf, lsystem_conf):
     """
     Creates an child (individual) through crossover with two parents
```
Comment (portaloffreedom) [inline pyrevolve/genotype/neat_brain_genome/crossover.py]: my fault, this should be "creates a child (genotype) through crossover with two parents"

File: pyrevolve/genotype/neat_brain_genome/neat_brain_genome.py [hunk 12]
```diff
@@ -188,19 +191,19 @@ def clone(self):
 
     @property
     def id(self):
-        return f"{self._neat_genome.GedID()}"
+        return f"{self._neat_genome.GetID()}"  #Change
```
Comment (portaloffreedom) [inline pyrevolve/genotype/neat_brain_genome/neat_brain_genome.py:194]: change?
GedID :man_facepalming: for myself...

File: pyrevolve/genotype/plasticoding/initialization.py [hunk 13]
```diff
@@ -58,3 +58,5 @@ def random_initialization(conf, next_robot_id):
     genotype.grammar = _generate_random_grammar(conf)
        
     return genotype
+
```
Comment (portaloffreedom) [inline pyrevolve/genotype/plasticoding/initialization.py]: useless extra spaces

File: pyrevolve/genotype/neat_brain_genome/crossover.py (hunk 10 above)
Comment (MRebolle) [reply pyrevolve/genotype/neat_brain_genome/crossover.py:24]: No parents here contain only the brain part of the genome

File: pyrevolve/genotype/neat_brain_genome/neat_brain_genome.py (hunk 12 above)
Comment (MRebolle) [reply pyrevolve/genotype/neat_brain_genome/neat_brain_genome.py:194]: #Change: So that I dont forget where I edited the original code in case I have to go back
